rest/oidc_api_test.go

```diff
@@ -1,108 +1,183 @@
 package rest
 
 import (
-	"io"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
 	"net/http"
+	"net/http/cookiejar"
 	"net/http/httptest"
 	"net/url"
-	"strings"
 	"testing"
+	"time"
 
 	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/db"
+	"github.com/gorilla/mux"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
 )
 
-func TestGetOIDCCallbackURL(t *testing.T) {
-	var (
-		wellKnownPath           = "/.well-known/openid-configuration"
-		headerLocation          = "Location"
-		authProvider            = "provider"
-		authRedirectURI         = "redirect_uri"
-		providerNameGoogle      = "google"
-		providerNameSalesforce  = "salesforce"
-		clientIDGoogle          = "client.id.google"
-		clientIDSalesforce      = "client.id.salesforce"
-		validationKeySalesforce = "validation.key.salesforce"
-		validationKeyGoogle     = "validation.key.google"
-	)
-
-	// Mock OpenID Connect Provider Discovery Endpoints
-	var issuerGoogle, issuerSalesforce string
-	mux := http.NewServeMux()
-	mux.HandleFunc("/google"+wellKnownPath, func(res http.ResponseWriter, req *http.Request) {
-		res.Header().Set("Content-Type", "application/json")
-		body := `{
-			"issuer": "${issuer}", 
-			"authorization_endpoint": "https://accounts.google.com/o/oauth2/v2/auth", 
-        	"token_endpoint": "https://oauth2.googleapis.com/token",
-			"jwks_uri": "https://www.googleapis.com/oauth2/v3/certs", 
-			"id_token_signing_alg_values_supported": ["RS256"]
-		}`
-		_, err := io.WriteString(res, strings.ReplaceAll(body, "${issuer}", issuerGoogle))
-		require.NoError(t, err, "Failed to write response body for google provider")
-	})
-	mux.HandleFunc("/salesforce"+wellKnownPath, func(res http.ResponseWriter, req *http.Request) {
-		res.Header().Set("Content-Type", "application/json")
-		body := `{
-			"issuer": "${issuer}", 
-			"authorization_endpoint": "https://login.salesforce.com/services/oauth2/authorize",
-        	"token_endpoint": "https://login.salesforce.com/services/oauth2/token",
-			"jwks_uri": "https://login.salesforce.com/id/keys", 
-			"id_token_signing_alg_values_supported": ["RS256"]
-		}`
-		_, err := io.WriteString(res, strings.ReplaceAll(body, "${issuer}", issuerSalesforce))
-		require.NoError(t, err, "Failed to write response body for salesforce provider")
-	})
+var (
+	tokenTypeBearer       = "Bearer"
+	mockSyncGatewayURL    = ""
+	issuerGoogle          = ""
+	issuerFacebook        = ""
+	authProvider          = "provider"
+	providerNameGoogle    = "google"
+	providerNameFacebook  = "facebook"
+	clientIDGoogle        = "client.id.google"
+	clientIDFacebook      = "client.id.facebook"
+	validationKeyFacebook = "validation.key.facebook"
+	validationKeyGoogle   = "validation.key.google"
+	wellKnownPath         = "/.well-known/openid-configuration"
+
+	wantTokenResponse OIDCTokenResponse
+)
```

Comment:
Refactored the variables which are scoped globally, but of them namely issuerGoogle, issuerFacebook, wantTokenResponse and invalidateState still needs to keep locally since the mocked response from auth server relies on the the values of these variable.